File: app/src/main/java/dev/danielholmberg/improve/Adapters/VipImagesAdapter.java
package dev.danielholmberg.improve.Adapters;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.SortedList;

import java.util.ArrayList;

import dev.danielholmberg.improve.Models.VipImage;
import dev.danielholmberg.improve.R;
import dev.danielholmberg.improve.ViewHolders.VipImageViewHolder;

public class VipImagesAdapter extends RecyclerView.Adapter<VipImageViewHolder>{
    private static final String TAG = VipImagesAdapter.class.getSimpleName();

    private String noteId;
    private boolean thumbnail;
    private SortedList<VipImage> vipImages;
    private View vipImageView;
    private boolean editMode = false;

    public VipImagesAdapter(String noteId, boolean thumbnail) {
        this.noteId = noteId;
        this.thumbnail = thumbnail;

        this.vipImages = new SortedList<>(VipImage.class, new SortedList.Callback<VipImage>() {
            @Override
            public int compare(@NonNull VipImage o1, @NonNull VipImage o2) {
                return o1.getId().compareTo(o2.getId());
            }

            @Override
            public void onChanged(int position, int count) {
                notifyItemChanged(position);
            }

            @Override
            public boolean areContentsTheSame(@NonNull VipImage oldItem, @NonNull VipImage newItem) {
                return oldItem.getId().equals(newItem.getId());
            }

            @Override
            public boolean areItemsTheSame(@NonNull VipImage oldItem, @NonNull VipImage newItem) {
                return oldItem.getId().equals(newItem.getId());
            }

            @Override
            public void onInserted(int position, int count) {
                notifyItemInserted(position);
            }

            @Override
            public void onRemoved(int position, int count) {
                notifyItemRemoved(position);
            }

            @Override
            public void onMoved(int fromPosition, int toPosition) {
                notifyItemMoved(fromPosition, toPosition);
            }
        });
    }

    @NonNull
    @Override
    public VipImageViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if(thumbnail) {
            vipImageView = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_vip_image_thumbnail, parent, false);
        } else {
            vipImageView = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_vip_image_preview, parent, false);
        }

        return new VipImageViewHolder(parent.getContext(), vipImageView, noteId, this);

    }

    @Override
    public void onBindViewHolder(@NonNull final VipImageViewHolder holder, int position) {
        if(thumbnail) {
            holder.bindModelToThumbnailView(this.vipImages.get(position));
        } else {
            holder.bindModelToPreviewView(this.vipImages.get(position));
        }

        holder.setEditMode(this.editMode);
    }

    @Override
    public int getItemCount() {
        return vipImages.size();
    }

    public void add(VipImage vipImage) {
        Log.d(TAG, "Adding vipImage: " + vipImage.getId());
        this.vipImages.add(vipImage);
    }

    public void remove(VipImage vipImage) {
        Log.d(TAG, "Removing vipImage: " + vipImage.getId());
        this.vipImages.remove(vipImage);
    }

    public void addImages(ArrayList<VipImage> images) {
        vipImages.beginBatchedUpdates();
        for(VipImage vipImage: images) {
            vipImages.add(vipImage);
        }
        vipImages.endBatchedUpdates();
    }

    public ArrayList<VipImage> getImageList() {
        ArrayList<VipImage> vipImageCopy = new ArrayList<>();
        for(int i = 0; i < vipImages.size(); i++) {
            vipImageCopy.add(vipImages.get(i));
        }
        return vipImageCopy;
    }

    public void setEditMode(boolean editMode) {
        this.editMode = editMode;
        notifyDataSetChanged();
    }

}

File: app/src/main/java/dev/danielholmberg/improve/Adapters/VipImagesAdapter.kt
package dev.danielholmberg.improve.Adapters

import android.util.Log
import androidx.recyclerview.widget.RecyclerView
import dev.danielholmberg.improve.ViewHolders.VipImageViewHolder
import androidx.recyclerview.widget.SortedList
import dev.danielholmberg.improve.Models.VipImage
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View
import dev.danielholmberg.improve.R
import java.util.ArrayList

class VipImagesAdapter(private val noteId: String, private val thumbnail: Boolean) :
    RecyclerView.Adapter<VipImageViewHolder>() {

    private val vipImages: SortedList<VipImage> =
        SortedList(VipImage::class.java, object : SortedList.Callback<VipImage>() {
            override fun compare(o1: VipImage, o2: VipImage): Int {
                return o1.id!!.compareTo(o2.id!!)
            }

            override fun onChanged(position: Int, count: Int) {
                notifyItemChanged(position)
            }

            override fun areContentsTheSame(oldItem: VipImage, newItem: VipImage): Boolean {
                return oldItem.id == newItem.id
            }

            override fun areItemsTheSame(oldItem: VipImage, newItem: VipImage): Boolean {
                return oldItem.id == newItem.id
            }

            override fun onInserted(position: Int, count: Int) {
                notifyItemInserted(position)
            }

            override fun onRemoved(position: Int, count: Int) {
                notifyItemRemoved(position)
            }

            override fun onMoved(fromPosition: Int, toPosition: Int) {
                notifyItemMoved(fromPosition, toPosition)
            }
        })

    private var vipImageView: View? = null
    private var editMode = false

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VipImageViewHolder {
        vipImageView = if (thumbnail) {
            LayoutInflater.from(parent.context)
                .inflate(R.layout.item_vip_image_thumbnail, parent, false)
        } else {
            LayoutInflater.from(parent.context)
                .inflate(R.layout.item_vip_image_preview, parent, false)
        }
        return VipImageViewHolder(parent.context, vipImageView!!, noteId, this)
    }

    override fun onBindViewHolder(holder: VipImageViewHolder, position: Int) {
        if (thumbnail) {
            holder.bindModelToThumbnailView(vipImages[position])
        } else {
            holder.bindModelToPreviewView(vipImages[position])
        }
        holder.setEditMode(editMode)
    }

    override fun getItemCount(): Int {
        return vipImages.size()
    }

    fun add(vipImage: VipImage) {
        Log.d(TAG, "Adding vipImage: " + vipImage.id)
        vipImages.add(vipImage)
    }

    fun remove(vipImage: VipImage) {
        Log.d(TAG, "Removing vipImage: " + vipImage.id)
        vipImages.remove(vipImage)
    }

    fun addImages(images: ArrayList<VipImage>) {
        vipImages.beginBatchedUpdates()
        for (vipImage in images) {
            vipImages.add(vipImage)
        }
        vipImages.endBatchedUpdates()
    }

    val imageList: ArrayList<VipImage>
        get() {
            val vipImageCopy = ArrayList<VipImage>()
            for (i in 0 until vipImages.size()) {
                vipImageCopy.add(vipImages[i])
            }
            return vipImageCopy
        }

    fun setEditMode(editMode: Boolean) {
        this.editMode = editMode
        notifyDataSetChanged()
    }

    companion object {
        private val TAG = VipImagesAdapter::class.java.simpleName
    }
}